### DISCUSSION TITLE
Layout Utility

### DISCUSSION THREAD

**Discussion by bcakmakoglu** [bcakmakoglu/vue-flow#61] [Ideas]
I'm thinking of creating a built-in utility to provide a layout to the nodes, basically like the current example does using dagre. 
Dagre is not being maintained afaik and I haven't found any viable alternatives, so if any of you know one, I'd be happy to hear about it. 

The layouting would receive a collection of nodes that have no specific positions set (or we can just overweite their positions) and group them by levels which are then aligned either verically or horizontally. 

I will update the post with an example when I got one ready but I don't expect this feature to be available before 0.4.1.

If you are interested in this feature let me know by reacting to the post or commenting ☺
0.4.0 has priority over any new features currently, for obvious reasons.

**Comment by bianpratama** [comment]
@bcakmakoglu 

XState Vizualizer is using `elkjs` for their auto-layout algorithm.
- https://github.com/kieler/elkjs
- https://github.com/statelyai/xstate-viz/blob/4ce37950f942013c8e4ee16eeb67b24a9309862f/src/graphUtils.ts
- https://github.com/statelyai/xstate-viz/blob/44e2f3061244967bbc7f769ede4a419efcb5ad77/src/directedGraph.ts

**Reply by bcakmakoglu** [reply]
Thanks for the suggestion!
I will look into it; I was wondering what XState was using but didn't look into it 😅

**Reply by bianpratama** [reply]
Eventually, I'm going with `d3-force` + `d3-hierarchy`. So that, it aligns with how `vue-flow` uses d3 family: `d3-drag`, `d3-selection`,  and `d3-zoom`. And it will be easier to extend to other layouting algorithm.

Below is preview of the implementation, using [forced-directed tree](https://observablehq.com/@d3/force-directed-tree?collection=@d3/d3-force):
![Implementation Preview](https://ucarecdn.com/bab2794c-c114-43c8-9bc4-09544b003e34/)

**Reply by bcakmakoglu** [reply]
Yeah React Flow is using a similar concept for their forced layout examples, so it's not a bad idea to look into what's available with d3.

**Reply by bianpratama** [reply]
Yes, I initially got the idea from React Flow's example of [Force Layout](https://reactflow.dev/docs/examples/layout/force-layout/). But because it's behind a paywall, I tried to play with my own implementation using Vue Flow.

I will try to contribute a PR for Vue Flow's examples, when my implementation have been more stabilized.

**Reply by bcakmakoglu** [reply]
That would be great! I am looking forward to it 🚀

**Comment by JonasPardon** [comment]
Hey there!

I'm exploring Vue Flow as a replacement for mermaidjs for a package I'm developing that shows a flowchart of an event-driven application. I wanted to make it more interactive and landed on this package for now as it looks very sleek 👌🏻 

The only issue so far is the fact that I currently let mermaidjs do the layout and calculating it myself is a bit complex. 

My question is: do you think auto-layout would support something along the lines of the attached example, and if you think it's feasible, are you still planning a layout utility any time soon?

Either way, thanks for the package, I'm for sure keeping this one in the back of my mind for non-pet-projects as well!

<img width="771" alt="image" src="https://user-images.githubusercontent.com/10206461/211665976-0daffb2e-c5ef-4881-8fe9-b9836424fa8d.png">

**Reply by bcakmakoglu** [reply]
There are layouting algorithms that handle these cases, like dagre, elkjs, d3-hierarchy.
Have you tried checking those out and implementing them on top of your graph for layouting? :)
I'm still not sure if I will be adding a layout utility - so until something like that happens, you might want to arrange yourself differently.

**Reply by JonasPardon** [reply]
Thanks for the tip, I'll give it a shot!

**Reply by bcakmakoglu** [reply]
An example using dagre can be found [here](https://github.com/bcakmakoglu/vue-flow/tree/master/examples/vite/src/Layouting), you can use that as a starting point :)
